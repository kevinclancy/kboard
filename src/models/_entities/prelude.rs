//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.13

pub use super::boards::Entity as Boards;
pub use super::replies::Entity as Replies;
pub use super::reply_statuses::Entity as ReplyStatuses;
pub use super::threads::Entity as Threads;
pub use super::users::Entity as Users;
