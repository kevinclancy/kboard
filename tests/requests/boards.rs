use insta::{assert_debug_snapshot};
use kboard::{app::App, models::users};
use loco_rs::testing::prelude::*;
use serial_test::serial;

// TODO: see how to dedup / extract this to app-local test utils
// not to framework, because that would require a runtime dep on insta
macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("board_request");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_get_boards() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        seed::<App>(&ctx).await.unwrap();
        let response = request.get("/api/boards").await;
        assert_eq!(
            response.status_code(),
            200,
            "Boards request should succeed"
        );

        assert_debug_snapshot!(response.text());
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_get_threads() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        seed::<App>(&ctx).await.unwrap();
        let response = request.get("/api/boards/1/threads").await;
        assert_eq!(
            response.status_code(),
            200,
            "Threads request should succeed"
        );

        assert_debug_snapshot!(response.text());
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_post_thread() {
    configure_insta!();

    let payload = serde_json::json!({
        "title": "I was startled by a Pika.",
        "initial_reply_text": "The Pika's cry caught me off guard.",
    });

    request::<App, _, _>(|request, ctx| async move {
        seed::<App>(&ctx).await.unwrap();

        // Get user1 from fixtures and generate JWT token
        let user1 = users::Model::find_by_pid(&ctx.db, "11111111-1111-1111-1111-111111111111").await.unwrap();
        let jwt_secret = ctx.config.get_jwt_config().unwrap();
        let token = user1
            .generate_jwt(&jwt_secret.secret, jwt_secret.expiration)
            .unwrap();

        let response = request
            .post("/api/boards/1/threads")
            .add_header("authorization", format!("Bearer {}", token))
            .json(&payload)
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        assert_debug_snapshot!(response_text);

        // now read the new thread and confirm it exists
        let response = request.get("/api/boards/1/threads/5/replies?page_size=1&page_number=0").await;
        assert_eq!(response.status_code(), 200);
    })
    .await;
}

