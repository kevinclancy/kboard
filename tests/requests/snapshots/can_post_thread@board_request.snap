---
source: tests/requests/boards.rs
expression: response.text()
---
"{\"thread_id\":5}"
